### PR TITLE
Breaking: Strict package exports (refs #13654)

### PR DESCRIPTION
**Pull request by nzakas**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/master/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[x] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

This implements the [strict package exports RFC](https://github.com/eslint/rfcs/tree/main/designs/2021-package-exports) by updating `api.js` and implementing `unsupported-api.js`.

This does **not** implement `ESLint#getRulesMetaForReport()` method. I'll do a separate PR for that piece.

#### Is there anything you'd like reviewers to focus on?

Please check that this matches the RFC. Also, how do we want to document the `use-at-your-own-risk` entrypoint? Should that be on the Node.js API page, or should we just omit it and trust that people who want it will find it?

### CODE REVIEW

**Comment by nzakas** [comment]
Hmmm, there's a ton of linting errors on the master branch right now.

**Comment by btmills** [comment]
I'm not able to repro any failures on `master` locally, and the last commit to `master` passed checks: https://github.com/eslint/eslint/runs/2809970160. Your first commit changed a bunch of rule tests, which I wasn't expecting. Were those changes intentional?

**Comment by mdjermanovic** [comment]
You probably have an old version of `eslint-plugin-eslint-plugin` locally installed.

**Comment by nzakas** [comment]
Ah ok, weird. I definitely didn’t intend to change all of those files it must have happened when I ran auto fix. I’ll try updating the plugin

**Comment by nzakas** [comment]
Thanks @mdjermanovic that did the trick!

**Comment by nzakas** [comment]
Forgot to mention: the more I think about it, the more I feel like the `use-at-your-own-risk` entrypoint should not be documented. People were already digging around the `eslint` package and finding APIs to use that weren't documented before. I don't want to implicitly "bless" stuff in `use-at-your-own-risk` by documenting it, but would like to hear other thoughts.

**Comment by ljharb** [comment]
By documented, you mean in prose? as long as it's in package.json "exports" that seems fine.

**Comment by nzakas** [comment]
@ljharb yes, in prose. Forgot people could just look in package.json 🤦

File: docs/developer-guide/nodejs-api.md [hunk 1]
```diff
@@ -30,8 +30,6 @@ While ESLint is designed to be run on the command line, it's possible to use ESL
     * [getRules()](#lintergetrules)
     * [defineParser()](#linterdefineparser)
     * [version](#linterversionlinterversion)
-* [linter (deprecated)](#linter-1)
-* [CLIEngine (deprecated)](#cliengine)
 * [RuleTester](#ruletester)
     * [Customizing RuleTester](#customizing-ruletester)
 * [Deprecated APIs](#deprecated-apis)
```
Comment (mdjermanovic) [inline docs/developer-guide/nodejs-api.md:35]: "Deprecated APIs" section is also removed

**Comment by mdjermanovic** [comment]
> Forgot to mention: the more I think about it, the more I feel like the `use-at-your-own-risk` entrypoint should not be documented. People were already digging around the `eslint` package and finding APIs to use that weren't documented before. I don't want to implicitly "bless" stuff in `use-at-your-own-risk` by documenting it, but would like to hear other thoughts.

I agree that we shouldn't document the `use-at-your-own-risk` entrypoint, as we don't want to encourage its use.

btw it seems that `exports` in package.json doesn't restrict `require(<absolute path>)`, so all internal modules still can be loaded from other packages, like this:

```js
const accessorPairs = require(
  path.resolve(path.dirname(require.resolve("eslint")), "rules", "accessor-pairs")
);
```

File: docs/developer-guide/nodejs-api.md (hunk 1 above)
Comment (mdjermanovic) [reply docs/developer-guide/nodejs-api.md:35]:
```suggestion
```

File: docs/developer-guide/nodejs-api.md [hunk 2]
```diff
@@ -460,8 +458,8 @@ Those run on `linter2` will get `process.cwd()` if the global `process` object i
 The most important method on `Linter` is `verify()`, which initiates linting of the given text. This method accepts three arguments:
 
 * `code` - the source code to lint (a string or instance of `SourceCode`).
-* `config` - a configuration object that has been processed and normalized by CLIEngine using eslintrc files and/or other configuration arguments.
-    * **Note**: If you want to lint text and have your configuration be read and processed, use CLIEngine's [`executeOnFiles`](#cliengineexecuteonfiles) or [`executeOnText`](#cliengineexecuteontext) instead.
+* `config` - a configuration object that has been processed and normalized by `ESLint`` using eslintrc files and/or other configuration arguments.
+    * **Note**: If you want to lint text and have your configuration be read and processed, use [`ESLint#lintFiles()`](#eslint-lintfiles) or [`ESLint#lintText()`](#eslint-linttext) instead.
```
Comment (mdjermanovic) [inline docs/developer-guide/nodejs-api.md:462]:
```suggestion
* `config` - a configuration object that has been processed and normalized by `ESLint` using eslintrc files and/or other configuration arguments.
    * **Note**: If you want to lint text and have your configuration be read and processed, use [`ESLint#lintFiles()`][eslint-lintfiles] or [`ESLint#lintText()`][eslint-linttext] instead.
```

**Review by mdjermanovic — APPROVED** [review]
LGTM aside from two small documentation suggestions I left.

File: docs/developer-guide/nodejs-api.md (hunk 1 above)
Comment (nzakas) [reply docs/developer-guide/nodejs-api.md:35]: Thanks!

File: docs/developer-guide/nodejs-api.md (hunk 2 above)
Comment (nzakas) [reply docs/developer-guide/nodejs-api.md:461]: Thanks

**Comment by nzakas** [comment]
Thanks for catching those docs issues.

**Review by mdjermanovic — APPROVED** [review]

**Review by btmills — APPROVED** [review]

**Review by aladdin-add — APPROVED** [review]

**Review by snitin315 — APPROVED** [review]

**Pull request merged**